core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -167,8 +167,20 @@ class LogCleaner(initialConfig: CleanerConfig,
    */
   def shutdown(): Unit = {
     info("Shutting down the log cleaner.")
-    cleaners.foreach(_.shutdown())
-    cleaners.clear()
+    try {
+      cleaners.foreach(_.shutdown())
+      cleaners.clear()
+    } finally {
+      remoteMetrics()
+    }
+  }
+
+  def remoteMetrics(): Unit = {
+    metricsGroup.removeMetric("max-buffer-utilization-percent")
+    metricsGroup.removeMetric("cleaner-recopy-percent")
+    metricsGroup.removeMetric("max-clean-time-secs")
+    metricsGroup.removeMetric("max-compaction-delay-secs")
+    metricsGroup.removeMetric("DeadThreadCount")
```

Comment:
I had a similar thought. Calling them out together as constants could also remind future contributors that might add a metric that they need to remove the metric too.

Reply:
Good suggestion both. I have fixed this in recent commit.

> Just for my own understanding since I am a newbie: what happens when a metric is removed, concretely, what happens to the already registered value?

@machi1990, it's a longer conversation because the metrics story in Kafka is a bit complicated, perhaps for another time. The summary is that we use `Metrics-Core` (or rather, an ancient version of it). There is a central registry where it stores all metric names. the registry is initialised once per server. Every time, we "add"/"remove" a metric, we add/remove to that registry. In addition to this Kafka also supports exposing metrics through JMX MBeans. When we add/remove a metrics, we add/remove the registered beans to the JMX server.

You can find the metrics-core concepts such as registry explained here: https://metrics.dropwizard.io/4.2.0/manual/core.html

Other stuff where you can read about out metrics story: 
1. https://cwiki.apache.org/confluence/display/KAFKA/KIP-510%3A+Metrics+library+upgrade
2. Ticket to move core to new metrics library: https://issues.apache.org/jira/browse/KAFKA-1930
3. Ticket to upgrade metrics-core to dropwizard: https://issues.apache.org/jira/browse/KAFKA-960

Reply:
Thanks @divijvaidya the change looks good to me. 
Also thanks for the reply regarding my other question. 
I noticed an occurrence of this in other classes that do not remove the metric during shutdown: e.g `ClientQuotaManager.scala` and some of its child classes. Is it something that you've observed as well? I could open a JIRA so that we track it and work on it in separate PR